text/0300-programmatic-toolkit.md

```diff
@@ -0,0 +1,745 @@
+
+```ts
+const io = new CdkCliHost({
+  isInteractive: true
+})
+```
+
+##### SdkProvider
+
+  [Alternative: Make this `CredentialsProvider` or `SdkConfigProvider` instead]
+
+The `AwsSdkProvider` defines how the toolkit interacts with AWS services.
+A default provider is available.
+You can provide a custom class implementing the `IAwsSdkProvider` interface.
+A standard `SdkProviderForCdk` is available that includes all features of the AWS CDK CLI [TODO: this might end up in a separate package].
+
+#### Cloud Executables
+
+A Cloud Executable is the primitive that defines a deployable AWS CDK application.
+AWS CDK apps might need to be synthesized multiple times with additional context values before they are ready.
+This concept is what the Cloud Executable represents.
+
+The most flexible way to obtain a Cloud Executable is from a function returning a `CloudAssembly`.
+Providing a function allows your implementation to integrate with the [Programmatic Toolkit]'s multi-pass requirement.
+The function will be called one or more times with the known context values of the current iteration.
+It must then use these values to synthesize and return a `CloudAssembly`.
+For all features (e.g. lookups) to work correctly, the `App()` must be instantiated with the received context values.
+Since it is not possible to update the context of an app, it must be created inside the function scope.
+
+A basic implementation would look like this:
+
+```ts
+const cx: ICloudExecutable = CloudExecutable.fromCloudAssemblyProducer(
+  async (context: Record<string, any>): cxapi.CloudAssembly => {
+    const app = new cdk.App({ context });
+    const stack = new cdk.Stack(app);
+    return app.synth();
+  }
+);
+```
+
+Existing applications, can also be used by simply passing an AWS CDK app directory (containing a `cdk.json` file):
+
+```ts
+const cx: ICloudExecutable = CloudExecutable.fromCdkApp("/path/to/cdk/app");
+```
+
+Alternatively the `cdk.out` directory from a previous synth can be used as well:
+
+```ts
+const cx: ICloudExecutable = CloudExecutable.fromCdkOut("/path/to/cdk.out");
+```
+
+#### Actions - synth
+
+The `synth` action takes a Cloud Executable and synthesizes it in to a `CloudAssembly`.
+
+```ts
+const cx: ICloudExecutable =  await cdk.synth(cx, { stacks: ['MyTestStack'] });
+```
+
+It returns a Cloud Executable that is cached on disk and can be used as input for other actions.
+
+#### Actions - bootstrap
+
+The `bootstrap` action takes in a list of environments and bootstraps each environment with the provided configuration:
+
+```ts
+const envs: string[] = cx.environments();
+await cdk.bootstrap(envs, { trust: ['000000000000', '111111111111'] });
+```
+
+It returns a list of stack deployments.
+
+#### Errors
+
+All errors are thrown as exceptions and the integrator must be prepared to handle them.
+Expected errors are structured data and extend the `IoMessage` interface.
+
+```ts
+interface ToolkitError extends IoMessage {
```

Comment:
I would expect

```ts
class ToolkitError extends Error { ... }
```

What's the value in extending `IoMessage`?

Do we really need `level`, `time`, `command` ... ? (`code` of course is very useful)

Reply:
Make it easier for integrators. `level` can be argued. `time` is free, `command` I think is important for provenance and if we do the setup right also pretty much free.

Reply:
But this also works

```ts
class ToolkitError extends Error implements IoMessage { ... }
```